src/adapter/src/coord/sequencer/inner/peek.rs

```diff
@@ -698,10 +697,25 @@ impl Coordinator {
                         explain_ctx,
                     })
                 } else {
+                    // If our timestamp determination relies on a timestamp, and we're not already in a
+                    // transaction, then we need to acquire read holds.
+                    let read_holds = match timestamp_context.timestamp() {
+                        Some(read_ts)
+                            if !ctx
+                                .session()
+                                .transaction()
+                                .is_in_multi_statement_transaction() =>
+                        {
+                            self.acquire_read_holds(*read_ts, &id_bundle)
+                        }
+                        _ => ReadHolds::new(),
+                    };
+
                     PeekStage::Finish(PeekStageFinish {
                         validity,
                         plan,
                         id_bundle,
+                        read_holds,
```

Comment:
Rather then passing this through `PeekStageFinish` maybe you can just call

```rust
self.acquire_read_holds_auto_cleanup(session, timestamp, &id_bundle;
```

as @ggevay did in this PR: #24443.

Reply:
Good callout, I updated the PR to do this!